### PR TITLE
migration functions

### PR DESCRIPTION
**Pull request by BarelUklis**
created migration functions to build simple and complex database

### CODE REVIEW

File: src/store/coin-store/index.ts [hunk 1]
```diff
@@ -18,7 +18,7 @@ class CoinStore {
     this.currentSocialMediaCoin = coinSocialInfo;
   }
 
-  setSocialTwitterData(twitterData:ISocailTwitterData|undefined = undefined) {
```
Comment (YuvalKrispin) [inline src/store/coin-store/index.ts:21]: twitterData?:ISocailDataObject
change to this maybe

File: src/interfaces/user-data.interface.ts [hunk 2]
```diff
@@ -7,7 +7,7 @@
 //   whales?: number
 // }
 
-import { ITweetPost } from "../../functions/src/interfaces/user-tweets-interface";
+import { ITweetPost } from "../../functions/src/interfaces/tweets-info-interface";
```
Comment (YuvalKrispin) [inline src/interfaces/user-data.interface.ts:10]: tweets-info.interface file name

File: functions/src/socialAPIs/reddit/reddit-controller.ts [hunk 3]
```diff
@@ -4,20 +4,18 @@ import { addSocialDataToDB, setBasicInfo, socailObject } from "../utils";
...
 
 export const getRedditPostsApi = async (data: IDataFromFrontend) => {    
-    const socialInfoFromCoinGecko = await getCoinSocialMedias(data.name);
+    const coinData = await getCoinSocialMedias(data.name);
 
-    setBasicInfo(socialInfoFromCoinGecko);
+    setBasicInfo(coinData.name);
 
-    if (socialInfoFromCoinGecko.reddit) {
+    if (coinData.socialInfo.reddit) {
```
Comment (YuvalKrispin) [inline functions/src/socialAPIs/reddit/reddit-controller.ts:11]: coinData.socialInfo?.reddit

File: functions/src/socialAPIs/medium/medium-controller.ts [hunk 4]
```diff
@@ -5,15 +5,15 @@ import { addSocialDataToDB, setBasicInfo, socailObject } from "../utils";
 import { getMediumData } from "./get-medium-posts";
 
 export const getMediumPostsApi = async (data: IDataFromFrontend) => {
-    const socialInfoFromCoinGecko = await getCoinSocialMedias(data.name);
-    setBasicInfo(socialInfoFromCoinGecko);
-    if (socialInfoFromCoinGecko.medium) {
+    const coinData = await getCoinSocialMedias(data.name);
+    setBasicInfo(coinData.name);
+    if (coinData.socialInfo.medium) {
```
Comment (YuvalKrispin) [inline functions/src/socialAPIs/medium/medium-controller.ts:10]: coinData.socialInfo?.medium

File: functions/src/socialAPIs/medium/get-medium-posts.ts [hunk 5]
```diff
@@ -3,13 +3,15 @@ import { checkForKeyWords } from "../utils";
...
 
 export const getMediumData = async (mediumUrl: string): Promise<IMediumPost[]> => {
+  let type = 'after';
   let mediumUser = "";
   const mediumUrlSplit = mediumUrl?.split("medium.com/");
   if (mediumUrlSplit) {
     const mediumUserWithoutAt = mediumUrlSplit[1].replace("@", "");
     mediumUser = mediumUserWithoutAt;
   }
   if (mediumUrlSplit && mediumUrlSplit[1] === "") {
```
Comment (YuvalKrispin) [inline functions/src/socialAPIs/medium/get-medium-posts.ts:13]: mediumUrlSplit && !mediumUrlSplit[1]

File: functions/src/socialAPIs/coins/coin-gecko.ts [hunk 6]
```diff
@@ -35,49 +63,50 @@ const getCoinSocialMedias = async (name: string): Promise<ISocailInfo>=> {
...
-      socailInfo.redditPostsTwoDays = response.data.community_data.reddit_average_posts_48h;
-      socailInfo.redditSubscribers = response.data.community_data.reddit_subscribers;
-      socailInfo.twitterFollowers = response.data.community_data.twitter_followers;
-      socailInfo.telegramUsersCount = response.data.community_data.telegram_channel_user_count;
+      socialInfo.redditPostsTwoDays = response.data.community_data.reddit_average_posts_48h;
+      socialInfo.redditSubscribers = response.data.community_data.reddit_subscribers;
+      socialInfo.twitterFollowers = response.data.community_data.twitter_followers;
+      socialInfo.telegramUsersCount = response.data.community_data.telegram_channel_user_count;
     }
     if (response.data.name) {
```
Comment (YuvalKrispin) [inline functions/src/socialAPIs/coins/coin-gecko.ts:71]: add to all if "?"

File: functions/src/migrations.ts [hunk 7]
```diff
@@ -1,21 +1,376 @@
...
-      .collection('coinsForUi')
-      .doc('coins')
-      .set({ coins: coinsToDb });
-  });
\ No newline at end of file
+      .collection('coins-data')
+      .doc(firstQuarterData[i].name)
+      .set(firstQuarterData[i]);
+
+    if (i % 29 === 0) {
```
Comment (YuvalKrispin) [inline functions/src/migrations.ts:370]: why?

File: functions/src/migrations.ts [hunk 8]
```diff
@@ -1,21 +1,376 @@
...
+      await new Promise(r => setTimeout(r, 65000));
```
Comment (YuvalKrispin) [inline functions/src/migrations.ts:371]: the file is too long 
split each functionality to autils file
try not to pass 100 lines

File: functions/src/migrations.ts [hunk 9]
```diff
@@ -1,21 +1,376 @@
...
+
+
+  const firstQuarterData = firstQuarter.data()?.coins;
+  const secondQuarterData = secondQuarter.data()?.coins;
+  const thirdQuarterData = thirdQuarter.data()?.coins;
+  const fourthQuarterData = fourthQuarter.data()?.coins;
+
+  const coinGeckoExtendedData = await getCoinGeckoCoinsExtendedDataListEndpoint();
+
+  for (let i = 0; i < firstQuarterData.length; i++) {
```
Comment (YuvalKrispin) [inline functions/src/migrations.ts:201]: foreach please

File: functions/src/migrations.ts [hunk 10]
```diff
@@ -1,21 +1,376 @@
...
+        lunarCrushId: coin.id,
+        image: coin.logo
+      });
+    } else {
+      finalArray[index].lunarCrushId = coin.id;
+      finalArray[index].image = coin.logo;
+    }
+  });
+
+  // const coinsDataFromCoinGecko = await getCoinGeckoCoinExtendedDataEndpoint();
```
Comment (YuvalKrispin) [inline functions/src/migrations.ts:137]: remove if no need

File: functions/src/migrations.ts [hunk 11]
```diff
@@ -1,21 +1,376 @@
...
+          thumb: coinGeckoExtendedData.data.image.thumb,
+          small: coinGeckoExtendedData.data.image.small,
+          large: coinGeckoExtendedData.data.image.large,
+        }
+      }
+      if(coinGeckoExtendedData.data.description.en !== "") {
+        // clean all html tags and r 
+        coin.description = coinGeckoExtendedData.data.description.en.replace(/(<([^>]+)>)/gi, "").replace(/(\r\n|\n|\r)/gm, "");
+      }
+      if(coinGeckoExtendedData.data.country_origin !== "") {
```
Comment (YuvalKrispin) [inline functions/src/migrations.ts:78]: if(coinGeckoExtendedData.data.country_origin) is the same as you did less test

File: functions/src/migrations.ts [hunk 12]
```diff
@@ -1,21 +1,376 @@
...
+        coin.social = socailInfo;        
+      }
+      if(coinGeckoExtendedData.data.image) {
+        coin.image = {
+          thumb: coinGeckoExtendedData.data.image.thumb,
+          small: coinGeckoExtendedData.data.image.small,
+          large: coinGeckoExtendedData.data.image.large,
+        }
+      }
+      if(coinGeckoExtendedData.data.description.en !== "") {
```
Comment (YuvalKrispin) [inline functions/src/migrations.ts:74]: this `coinGeckoExtendedData.data.description.en !== ""`
is the same as `coinGeckoExtendedData.data.description.en`

File: functions/src/migrations.ts [hunk 13]
```diff
@@ -1,21 +1,376 @@
...
+          socailInfo.homepage = links.homepage[0];
+        }
+        if (links.repos_url.github[0] !== "") {
+          socailInfo.github = links.repos_url.github[0];
+        }
+        // remove all falsy values
+        Object.keys(socailInfo).forEach(key => socailInfo[key] || delete socailInfo[key]);
+        coin.social = socailInfo;        
+      }
+      if(coinGeckoExtendedData.data.image) {
```
Comment (YuvalKrispin) [inline functions/src/migrations.ts:67]: `coinGeckoExtendedData.data?.image`

File: functions/src/migrations.ts [hunk 14]
```diff
@@ -1,21 +1,376 @@
...
+            socailInfo.medium = links.official_forum_url && checkIfIncludesMedium(links.official_forum_url);
+          }
+        }
+        if (links.chat_url !== "") {
+          socailInfo.discord = links.chat_url && checkIfIncludesDiscord(links.chat_url);
+        }
+        if (links.homepage[0] !== "") {
+          socailInfo.homepage = links.homepage[0];
+        }
+        if (links.repos_url.github[0] !== "") {
```
Comment (YuvalKrispin) [inline functions/src/migrations.ts:60]: remove all !== "" 
no need for that

File: functions/src/migrations.ts [hunk 15]
```diff
@@ -1,21 +1,376 @@
...
+      symbol: coin.symbol,
+    }
+    await admin.firestore().collection('coins').doc(coin.name).set(coinToDb);
+  })
+};
+
+export const setCoinGeckoCoinsSocialDataToDBApi = async () => {
+  const coins = await admin.firestore().collection('coins').get();
+  const coinsData = coins.docs.map(doc => doc.data());
+  for(let i = 1; i < coinsData.length; i++) {
```
Comment (YuvalKrispin) [inline functions/src/migrations.ts:24]: use foreach

File: functions/src/migrations.ts (hunk 7 above)
Comment (BarelUklis) [reply functions/src/migrations.ts:370]: because this is the max request per minute

**Pull request merged**